src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h

```diff
@@ -0,0 +1,1637 @@
+  char *buf = nullptr;
+
+  using L = absl::container_internal::Layout<ceph_le32, MetaInt, omap_inner_key_le_t>;
+  static constexpr L layout{1, 1, 1}; // = L::Partial(1, 1, 1);
+
+public:
+  template <bool is_const>
+  struct iter_t {
+    friend class StringKVInnerNodeLayout;
+
+    template <typename iterator, typename const_iterator>
+    friend void copy_from_foreign(iterator, const_iterator, const_iterator);
+    template <typename iterator>
+    friend void copy_from_local(unsigned, iterator, iterator, iterator);
+
+    using parent_t = typename crimson::common::maybe_const_t<StringKVInnerNodeLayout, is_const>::type;
+
+    parent_t node;
+    uint16_t index;
+
+    iter_t(
+      parent_t parent,
+      uint16_t index) : node(parent), index(index) {}
+
+    iter_t(const iter_t &) = default;
+    iter_t(iter_t &&) = default;
+    iter_t &operator=(const iter_t &) = default;
+    iter_t &operator=(iter_t &&) = default;
+
+    operator iter_t<!is_const>() const {
+      static_assert(!is_const);
+      return iter_t<!is_const>(node, index);
+    }
+
+    iter_t &operator*() { return *this; }
+    iter_t *operator->() { return this; }
+
+    iter_t operator++(int) {
+      auto ret = *this;
+      ++index;
+      return ret;
+    }
+
+    iter_t &operator++() {
+      ++index;
+      return *this;
+    }
+
+    uint16_t operator-(const iter_t &rhs) const {
+      assert(rhs.node == node);
+      return index - rhs.index;
+    }
+
+    iter_t operator+(uint16_t off) const {
+      return iter_t(
+	           node,
+	           index + off);
+    }
+    iter_t operator-(uint16_t off) const {
+      return iter_t(
+	           node,
+	           index - off);
+    }
+
+    uint16_t operator<(const iter_t &rhs) const {
+      assert(rhs.node == node);
+      return index < rhs.index;
+    }
+
+    uint16_t operator>(const iter_t &rhs) const {
+      assert(rhs.node == node);
+      return index > rhs.index;
+    }
+
+    bool operator==(const iter_t &rhs) const {
+      assert(node == rhs.node);
+      return rhs.index == index;
+    }
+
+    bool operator!=(const iter_t &rhs) const {
+      return !(*this == rhs);
+    }
+
+    omap_inner_key_t get_node_key() const {
+      omap_inner_key_le_t kint = node->get_node_key_ptr()[index];
+      return omap_inner_key_t(kint);
+    }
+
+    char *get_node_val_ptr() {
+      auto tail = node->buf + BLOCK_SIZE;
+      if (*this == node->iter_end())
+        return tail;
+      else {
+        return tail - static_cast<uint32_t>(get_node_key().key_off);
+      }
+    }
+
+    const char *get_node_val_ptr() const {
+      auto tail = node->buf + BLOCK_SIZE;
+      if ( *this == node->iter_end())
+        return tail;
+      else {
+        return tail - static_cast<uint32_t>(get_node_key().key_off);
+      }
+    }
+
+    void set_node_val(std::string_view val) {
+      static_assert(!is_const);
+      std::strcpy((char*)get_node_val_ptr(), std::string(val).c_str()); //copy char* to char* include "\0"
```

Comment:
If this is a frequent operation, might want to verify performance against alternatives.

Reply:
any insert operation will set node val.

Reply:
Then, I strongly suggest changing this code. You are doing a lot of extra work by copying the original incoming 'val' into a new null-terminated string (or not null terminated, but then c_str() will perform a copy).
See a good discussion in https://quuxplusone.github.io/blog/2020/03/20/c-str-correctness/ (by 
Arthur O’Dwyer, who is a C++ expert).
See also the code produced in https://godbolt.org/z/K76fxb

Reply:
do you mean we change it back to use const string& as input instead of string_view here?

Reply:
That's one (good) option.

Reply:
It seems I have to change each caller in call stack input string_view to const string& since string_view can't be convert to const string& directly.  do you agree it?

Reply:
Yes, it seems so.